data/tokenizer.py

```diff
@@ -120,3 +126,54 @@ def untokenize(self, tokens: list[str]) -> pd.DataFrame:
         df = pd.DataFrame(velocities, columns=["velocity"])
 
         return df
+
+
+class DstartEncoder(MidiEncoder):
+    def __init__(self, bins: int = 200):
+        self.specials = ["<s>", "</s>", "<blank>"]
+        self.bins = bins
+        # Make a copy of special tokens ...
+        self.vocab = list(self.specials)
+        self.bin_edges = self.load_bin_edges()
+        # ... and add velocity tokens
+        self._build_vocab()
+        self.token_to_id = {token: it for it, token in enumerate(self.vocab)}
+
+    def load_bin_edges(self):
+        artifacts_path = to_absolute_path("artifacts/bin_edges.yaml")
+        with open(artifacts_path, "r") as f:
+            bin_edges = yaml.safe_load(f)
+
+        dstart_bin_edges = bin_edges["dstart"][self.bins]
+        return dstart_bin_edges
+
+    def _build_vocab(self):
+        self.vocab += [str(possible_bin) for possible_bin in range(self.bins)]
+
+    @property
+    def vocab_size(self) -> int:
+        return len(self.vocab)
+
+    def tokenize(self, record: dict) -> list[str]:
+        # TODO I don't love the idea of adding tokens durint *tokenize* call
+        # If we want to pretend that our midi sequences have start and finish
+        # we should take care of that before we get here :alarm:
+
+        df = pd.DataFrame(record)
+
+        # Quantize only dstart
+        df["next_start"] = df.start.shift(-1)
+        df["dstart"] = df.next_start - df.start
+        df["dstart_bin"] = np.digitize(df.dstart.fillna(0), self.bin_edges) - 1
+
+        # get tokens from quantized data
+        tokens = [str(dstart_bin) for dstart_bin in df["dstart_bin"]]
+
+        tokens = ["<s>"] + tokens + ["</s>"]
+        return tokens
```

Comment:
I decided to simply make quantization inside tokenize method, which does seem very appropriate but also looks clean and works well with our architecture.

Reply:
I see the temptation of joining quantization with tokenization - maybe a single `Encoder` class could manage all of this in principle. But here, I'm also worried about the performance - I'm guessing that creating a new data frame for each sequence is going to slow down training a lot, and I don't want to cache data after tokenization.

Another problem here is that this function now does to separate steps, so if we wanted to keep this approach, it should be two functions `quantize` and `tokenize`.

Reply:
Another approach I think, would be to apend "tgt_bins" key to our dataset. We would have to make the quantization back in our build_translation_dataset function and include it in quantized_piece_to_records sequence. 
We could then use it to handle dstart as well as duration, but it is redundant when using velocity (which is "quantized" into 128 bins :) ), but if we wanted to predict both dstart and duration at some point we might want to find a better way...
Or we can just create completly seperate program for dstart prediction but I do not think this is such a good idea.